docs/SOFTWARE_DESIGN.md

```diff
@@ -0,0 +1,124 @@
+
+## Technical Architecture
+
+The game engine back-end will be written in C++ and the UI front-end will be written in browser languages, HTML5/JavaScript/CSS. Optionally, we could use a framework such as React or TypeScript for the front-end, but since the application will be small, plain HTML/CSS/JS without a build stage may be simplest.
+
+Unit testing will be performed on the C++ game engine code using [`utest.h`](https://github.com/sheredom/utest.h).
+
+The browser user interface will be tested with [Jest](https://github.com/facebook/jest) and [Puppeteer](https://github.com/puppeteer/puppeteer/).
+
+The turn-based nature of puzzle games and the HTML user interface makes it very easy to test our code, as opposed to a live action game on a screen where it'd be very difficult to programmatically provide input and assert on game state. Real-time graphics-based games can be math-heavy and asset-heavy.
+
+There is a very clear separation between front- and back-ends, so we can delegate disjoint tasks and responsibilities. The back-end game logic can be written in C++ without any knowledge of JS (and could ideally be modular enough to couple with a command-line front-end that might be useful for development). Greg will mostly be responsible for linking the JS and C++ components together.
+
+```
+           BACKEND (game engine)
+
+       .-------.
+       | uTest |
+       `-------`
+           |
+           v        [emscripten compilation]
+.-----------------.          |
+| C++ game engine |----------+--+
+`-----------------`             |
+                                |
+................................|.............
+                                |
+.------.   FRONTEND (browser)   |
+| User |                        |
+`------`                        |
+   |                            |
+[click events]    [API calls]   |
+   |                   |        v
+   |   .----------.~~~~+~~~>.------.
+   +-->| HTML5 UI |         | WASM |<--compiled backend engine
+       `----------`<~~~+~~~~`------`
+            ^          |
+            |     [game state]
+            |
+            |
+.----------------------.
+| Jest/Puppeteer tests |
+`----------------------`
+```
+
+Greg has built a simple proof-of-concept that illustrates the concept/workflow and relevant technologies interoperating: <https://github.com/ggorlen/wasm-click-counter>.
+
+### Game engine back-end
+
+The Sokoban game engine will consist of a top-level C++ class, `Sokoban`, which will expose the following interface:
+
+```
+// one of the possible objects that can exist on a cell
+enum Cell {
+   ' ': empty space
+   '#': impassable wall
+   '.': goal cell
+   '$': box
+   '*': box on top of goal cell
+   '@': player
+   '+': player on top of goal cell
+}
```

Comment:
An empty space corresponds to a cell within the walls, right. Are the outside of the wall will also be represented as '  '. I was thinking of adding a graphical component in the negative space of the map. This is maybe for future consideration.

Reply:
Yeah, that's a good point. It's been awhile, but I don't think the typical level format has such a character: https://www.sokoban-online.de/sokoban/levell-format/. Looking at it again, I'll probably change the representation a bit here to match that better in terms of UDLR for moves/directions.

In any case, this should be easy to handle in JS with something like `board.replace(/^ +/gm, "X")` where `"X"` is floor tile and `board` is a 2d string (or iterate the rows of an array if it's an array and `replace` on each one) that's outside the board, then we can render that using a separate tile.